cantools/database/can/conversion.py

```diff
@@ -0,0 +1,190 @@
+
+    def scaled_to_raw(self, scaled_value: SignalValueType) -> Union[int, float]:
+        if not isinstance(scaled_value, (int, float)):
+            raise TypeError(
+                f"'scaled_value' must have type 'int' or 'float' (is {type(scaled_value)})"
+            )
+        _raw = (scaled_value - self.offset) / self.scale
+        if self.is_float:
+            return _raw
+        return round(_raw)
+
+
+class NamedSignalConversion(BaseConversion):
+    is_float = False
+
+    def __init__(self, scale: float, offset: float, choices: Choices) -> None:
+        self.scale = scale
+        self.offset = offset
+        self._inverse_choices: Optional[Dict[str, int]] = None
+        self.choices = choices
+        self._update_choices()
+
+        self._conversion = conversion_factory(
+            scale=self.scale,
+            offset=self.offset,
+            choices=None,
+            is_float=False,
+        )
+
+    def raw_to_scaled(
+        self,
+        raw_value: Union[int, float],
+        decode_choices: bool = True,
+    ) -> SignalValueType:
+        if decode_choices:
+            try:
+                return self.choices[raw_value]  # type: ignore[index]
+            except (AttributeError, KeyError):
+                pass
+        return self._conversion.raw_to_scaled(raw_value, False)
+
+    def scaled_to_raw(self, scaled_value: SignalValueType) -> Union[int, float]:
+        if isinstance(scaled_value, (int, float)):
+            return self._conversion.scaled_to_raw(scaled_value)
+
+        if isinstance(scaled_value, (str, NamedSignalValue)):
+            return self.choice_to_number(str(scaled_value))
+
+        raise TypeError
+
+    def set_choices(self, choices: Optional[Choices]) -> None:
+        self.choices: Optional[Choices] = choices
+        self._update_choices()
+
+    def _update_choices(self) -> None:
+        # we simply assume that the choices are invertible
+        if self.choices:
+            self._inverse_choices = {str(x[1]): x[0] for x in self.choices.items()}
+        else:
+            self._inverse_choices = None
+
+    def choice_to_number(self, choice: Union[str, "NamedSignalValue"]) -> int:
+        if self._inverse_choices is None:
+            raise KeyError(f"Signal does not exhibit any choices")
+        return self._inverse_choices[str(choice)]
+
+
+def conversion_factory(
+    scale: float = 1,
+    offset: float = 0,
+    choices: Optional[Choices] = None,
+    is_float: bool = False,
+) -> BaseConversion:
+    if choices is None:
+        if scale == 1 and offset == 0:
+            return IdentityConversion(is_float=is_float)
+
+        if _is_integer(scale) and _is_integer(offset) and not is_float:
+            return LinearIntegerConversion(scale=int(scale), offset=int(offset))
+
+        return LinearConversion(scale=scale, offset=offset, is_float=is_float)
+
+    return NamedSignalConversion(scale=scale, offset=offset, choices=choices)
```

Comment:
hm. we possibly might need an `NamedSignalIntegerConversion` because it is possible to define scaling factor of 1 and an integer offset with some named values (alternatively, all conversion classes could either return integers or floats)

Reply:
NamedSignalConversion calls the factory internally, so it always uses the correct conversion anyway